libsql-sqlite3/src/vectorIndex.c

```diff
@@ -0,0 +1,462 @@
+    if( i == 0 ){
+      size = snprintf(pBuf, nBufSize, "%s", prefix);
+    }else {
+      size = snprintf(pBuf, nBufSize, ",%s%d", prefix, i);
+    }
+    if( size < 0 ){
+      return -1;
+    }
+    pBuf += size;
+    nBufSize -= size;
+  }
+  if( nBufSize <= 0 ){
+    return -1;
+  }
+  return 0;
+}
+
+/**************************************************************************
+** VectorInRow utilities
+****************************************************************************/
+
+sqlite3_value* vectorInRowKey(const VectorInRow *pVectorInRow, int iKey) {
+  assert( 0 <= iKey && iKey < pVectorInRow->nKeys );
+  return pVectorInRow->pKeyValues + iKey;
+}
+
+i64 vectorInRowLegacyId(const VectorInRow *pVectorInRow) {
+  if( pVectorInRow->nKeys == 1 && sqlite3_value_type(pVectorInRow->pKeyValues + 0) == SQLITE_INTEGER ){
+    return sqlite3_value_int64(pVectorInRow->pKeyValues);
+  }
+  return 0;
+}
+
+int vectorInRowTryGetRowid(const VectorInRow *pVectorInRow, u64 *nRowid) {
+  if( pVectorInRow->nKeys != 1 ){
+    return -1;
+  }
+  if( sqlite3_value_type(vectorInRowKey(pVectorInRow, 0)) != SQLITE_INTEGER ){
+    return -1;
+  }
+  *nRowid = sqlite3_value_int64(vectorInRowKey(pVectorInRow, 0));
+  return 0;
+}
+
+int vectorInRowPlaceholderRender(const VectorInRow *pVectorInRow, char *pBuf, int nBufSize) {
+  int i;
+  assert( pVectorInRow->nKeys > 0 );
+  if( nBufSize < 2 * pVectorInRow->nKeys ){
+    return -1;
+  }
+  for(i = 0; i < pVectorInRow->nKeys; i++){
+    *(pBuf++) = '?';
+    *(pBuf++) = ',';
+  }
+  *(pBuf - 1) = '\0';
+  return 0;
+}
+
+int vectorInRowAlloc(sqlite3 *db, const UnpackedRecord *pRecord, VectorInRow *pVectorInRow, char **pzErrMsg) {
+  int rc = SQLITE_OK;
+  int type, dims;
+  struct sqlite3_value *pVectorValue = pRecord->aMem + 0;
+  pVectorInRow->pKeyValues = pRecord->aMem + 1;
+  pVectorInRow->nKeys = pRecord->nField - 1;
+  pVectorInRow->pVector = NULL;
+
+  if( pVectorInRow->nKeys <= 0 ){
+    rc = SQLITE_ERROR;
+    goto out;  
+  }
+
+  if( sqlite3_value_type(pVectorValue)==SQLITE_NULL ){
+    rc = SQLITE_OK;
+    goto out;
+  }
+
+  if( detectVectorParameters(pVectorValue, VECTOR_TYPE_FLOAT32, &type, &dims, pzErrMsg) != 0 ){
+    rc = SQLITE_ERROR;
+    goto out;
+  }
+
+  pVectorInRow->pVector = vectorAlloc(type, dims);
+  if( pVectorInRow->pVector == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+
+  if( sqlite3_value_type(pVectorValue) == SQLITE_BLOB ){
+    vectorInitFromBlob(pVectorInRow->pVector, sqlite3_value_blob(pVectorValue), sqlite3_value_bytes(pVectorValue));
+  } else if( sqlite3_value_type(pVectorValue) == SQLITE_TEXT ){ 
+    // users can put strings (e.g. '[1,2,3]') in the table and we should process them correctly
+    if( vectorParse(pVectorValue, pVectorInRow->pVector, pzErrMsg) != 0 ){
+      rc = SQLITE_ERROR;
+      goto out;
+    }
+  }
+  rc = SQLITE_OK;
+out:
+  if( rc != SQLITE_OK ){
+    vectorFree(pVectorInRow->pVector);
```

Comment:
Is it ok to free this vector before it's allocated? Line 225 can jump here with rc != SQLITE_OK

Reply:
So can a failure in vectorAlloc

Reply:
Yes, this is bug. I guess I forgot ` && pVectorInRow->pVector != NULL` guard here :-(